consensus/istanbul/core/handler_test.go

```diff
@@ -464,6 +497,305 @@ func TestCore_handlerMsg(t *testing.T) {
+		t.Fatal(err)
+	}
+	if err := groupA.Start(); err != nil {
+		t.Fatal(err)
+	}
+	if err := groupB.Start(); err != nil {
+		t.Fatal(err)
+	}
+	defer proposer.Stop()
+	defer groupA.Stop()
+	defer groupB.Stop()
+
+	lastProposal, _ := mockBackend.LastProposal()
+	lastBlock := lastProposal.(*types.Block)
+	validators := mockBackend.Validators(lastBlock)
+	//fmt.Println("validator addrs = ", validators.SubList(lastBlock.Hash(), istCore.currentView()))
+
+	// proposer creates a block for group A
+	// proposer creates the other block for group B
+	{
+		msgSender := validators.GetProposer()
+		msgSenderKey := validatorKeyMap[msgSender.Address()]
+
+		proposalA, err := genBlockParams(lastBlock, msgSenderKey, 0, 0)
+		if err != nil {
+			t.Fatal(err)
+		}
+		proposalB, err := genBlockParams(lastBlock, msgSenderKey, 1000, 10)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		committeeSize := len(validators.SubList(lastBlock.Hash(), proposer.currentView()))
+		tmpList := validators.SubList(lastBlock.Hash(), proposer.currentView())
+
+		for i := range validators.SubList(lastBlock.Hash(), proposer.currentView()) {
+			// both groupA and groupB includes proposer as validator
+			if tmpList[i].Address() == msgSender.Address() {
+				tmpList = append(tmpList[i+1:], tmpList[:i]...)
+				break
+			}
+		}
+
+		// listA and ListB have the validator addresses of group A and group B
+		// the proposer address is included to each group
+		// each group makes consensus messages inside the group
+
+		listA := make([]istanbul.Validator, (committeeSize-1)/2)
+		listB := make([]istanbul.Validator, (committeeSize-1)/2)
+		copy(listA, tmpList[:(committeeSize-1)/2])
+		listA = append(listA, msgSender)
+
+		copy(listB, tmpList[(committeeSize-1)/2:])
+		listB = append(listB, msgSender)
+
+		sendMessages := func(state uint64, proposal *types.Block, CNList []istanbul.Validator, group *core) {
+			for _, val := range CNList {
+				v := validatorKeyMap[val.Address()]
+				if state == msgPreprepare {
+					istanbulMsg, _ := genIstanbulMsg(state, lastBlock.Hash(), proposal, msgSender.Address(), v)
+					err = group.handleMsg(istanbulMsg.Payload)
+				} else {
+					istanbulMsg, _ := genIstanbulMsg(state, lastBlock.Hash(), proposal, val.Address(), v)
+					err = group.handleMsg(istanbulMsg.Payload)
+				}
+				if err != nil {
+					fmt.Println(err)
+				}
+			}
+
+		}
+
+		// the proposer sends two different blocks to each group
+		sendMessages(msgPreprepare, proposalA, listA, groupA)
+		sendMessages(msgPreprepare, proposalB, listB, groupB)
+
+		// each group receives a block and process it by using handleMsg
+		// the proposer handles both messages to send consensus messages to both groups
+		sendMessages(msgPrepare, proposalA, listA, groupA)
+		sendMessages(msgCommit, proposalA, listA, groupA)
+
+		preparesSizeA := groupA.current.Prepares.Size()
+		commitsSizeA := groupA.current.Commits.Size()
+		fmt.Println("Group A: prepare size = ", preparesSizeA, " commit size = ", commitsSizeA)
+
```

Comment:
For each group, both groups need to be in committed state. Assert the following condition:
```
assert.True(t, groupA.state.Cmp(StateCommitted) == 0)
assert.True(t, groupB.state.Cmp(StateCommitted) == 0)
```

Reply:
reflected at 047a8c8